### PR TITLE
BUG: Fix #16363 - Prevent visit_BinOp from accessing value on UnaryOp

### LINKED ISSUE #16363: DataFrame.eval errors with AttributeError: 'UnaryOp'
#### Code Sample, a copy-pastable example if possible

```
def test_unary():
    df = pd.DataFrame({'x': np.array([0.11, 0], dtype=np.float32)})
    res = df.eval('(x > 0.1) | (x < -0.1)')
    assert np.array_equal(res, np.array([True, False])), res
```

#### Problem description

This is related to #11235. 
on python 3.6, pandas 20.1, this raises an error the traceback ends with:

```
  File ".../envs/py3/lib/python3.6/site-packages/pandas/core/computation/expr.py", line 370, in _maybe_downcast_constants
    name = self.env.add_tmp(np.float32(right.value))
AttributeError: 'UnaryOp' object has no attribute 'value'
```
In that case the right is -(0.1)

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Linux
OS-release: 4.8.0-49-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.20.1
pytest: None
pip: 9.0.1
setuptools: 27.2.0
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.19.0
xarray: None
IPython: 6.0.0
sphinx: None
patsy: None
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.0.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
pandas_gbq: None
pandas_datareader: None
</details>

Another example:

```python
>>> df = pd.DataFrame({'x':[1,2,3,4,5]})
>>> df.eval('x.shift(-1)')
```

### PR DESCRIPTION
**Pull request by alexcwatt**
- [x] closes #16363
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Checking `hasattr(side, 'value')` resolves the bug and is consistent with the way that other methods of the `BaseExprVisitor` class work (e.g., both `visit_Call_35` and `visit_Call_legacy` check for 'value').

### CODE REVIEW

File: pandas/tests/computation/test_eval.py [hunk 1]
```diff
@@ -608,6 +608,12 @@ def test_unary_in_array(self):
                       -False, False, ~False, +False,
                       -37, 37, ~37, +37], dtype=np.object_))
 
+    def test_float_comparison_bin_op(self):
+        # GH 16363
+        df = pd.DataFrame({'x': np.array([0], dtype=np.float32)})
+        res = df.eval('x < -0.1')
+        assert np.array_equal(res, np.array([False])), res
```
Comment (WillAyd) [inline pandas/tests/computation/test_eval.py]: What's the reason for `, res` here? I think also causing CI failure

File: pandas/tests/computation/test_eval.py [hunk 2]
```diff
@@ -608,6 +608,12 @@ def test_unary_in_array(self):
                       -False, False, ~False, +False,
                       -37, 37, ~37, +37], dtype=np.object_))
 
+    def test_float_comparison_bin_op(self):
+        # GH 16363
+        df = pd.DataFrame({'x': np.array([0], dtype=np.float32)})
```
Comment (WillAyd) [inline pandas/tests/computation/test_eval.py]: Is this specific to float32? Read briefly through issue and seemed like it affected other things

**Review by WillAyd — CHANGES_REQUESTED** [review]
Can you add a whatsnew note for v0.25?

File: pandas/tests/computation/test_eval.py (hunk 2 above)
Comment (alexcwatt) [reply pandas/tests/computation/test_eval.py]: @WillAyd The bug arises when one side returns a float32 and the other side is a scalar that doesn't have a 'value' attribute.

File: pandas/tests/computation/test_eval.py (hunk 1 above)
Comment (alexcwatt) [reply pandas/tests/computation/test_eval.py]: Good catch, removing this.

File: pandas/tests/computation/test_eval.py (hunk 2 above)
Comment (WillAyd) [reply pandas/tests/computation/test_eval.py]: Does this still work if the operands are switched? i.e. `-0.1 > x`?

File: doc/source/whatsnew/v0.25.0.rst [hunk 3]
```diff
@@ -228,6 +228,7 @@ Performance Improvements
 Bug Fixes
 ~~~~~~~~~
 
+- Bug in :meth:`~pandas.eval` when comparing floats with scalar operators that don't have a 'value' attribute (e.g., unary operators) (:issue:`25928`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: can you move down to numeric section. 

make this more like:

```
Bug in :meth:`~pandas.eval` when comparing floats with scalar operators, for example: ``x < 0.1`` (:issue:...)
```

File: pandas/tests/computation/test_eval.py (hunk 2 above)
Comment (jreback) [reply pandas/tests/computation/test_eval.py]: an you parameterize on the dtype (at least float32/float64)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/computation/test_eval.py (hunk 2 above)
Comment (alexcwatt) [reply pandas/tests/computation/test_eval.py]: > Does this still work if the operands are switched? i.e. `-0.1 > x`?

I will add a test with a negative value on the left.

Comment (alexcwatt) [reply pandas/tests/computation/test_eval.py]: > an you parameterize on the dtype (at least float32/float64)

Would you like a test for float64? This particular bug was caused by code that handles float32's, so float64's were fine, but I can add a test to ensure that this does not become an issue.

File: doc/source/whatsnew/v0.25.0.rst (hunk 3 above)
Comment (alexcwatt) [reply doc/source/whatsnew/v0.25.0.rst]: Yep, will do.

File: pandas/tests/computation/test_eval.py (hunk 2 above)
Comment (alexcwatt) [reply pandas/tests/computation/test_eval.py]: @jreback Oh, I see now pytest's `parametrize` in some of the other tests, I will update this

**Review by WillAyd — APPROVED** [review]

**Comment by alexcwatt** [comment]
@jreback Should be ready for another review.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @alexcwatt

**Comment by kenahoo** [comment]
@jreback & @alexcwatt - this was also an issue for string data, as I showed in this previous comment on #16363 :

https://github.com/pandas-dev/pandas/issues/16363#issuecomment-313231698

The test case is:

```
def test_unary():
    df = pd.DataFrame({'x': ["one", "two"]})
    df.eval('x.shift(-1)')
```

Is that also fixed now?  If not, we should keep #16363 open as not fixed.

**Comment by jreback** [comment]
you can test on master and if not pls open. new issue 

if it works then can add an additional test (pls do a PR)

**Comment by kenahoo** [comment]
I don't think I'll have time for that very soon, I'd recommend just re-opening #16363 until the full set of problems identified in the ticket can be verified as fixed.

**Comment by jreback** [comment]
plan make a new issue instead

**Pull request merged**